### PR TITLE
feat: Add Unit Tests for Mempool Functionality

### PR DESCRIPTION
**Pull request by eugypalu**
Resolves: #1355 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing

# What is the new behavior?

mempool unit tests

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by tcoratger** [comment]
@eugypalu Nice addition to our test suite. I would have two small remarks before merging:
- I think we will wait to merge https://github.com/kkrt-labs/kakarot-rpc/pull/1352 first because there will be a small change in the provider but normally it will just require to modify very slightly the way you call the pool which should be in the ethclient and no longer in the ethprovider (but it is not a big change for your tests).
- On the other hand, I know that it is rather recommended to separate the tests to have unit test as much as possible and limit the scope of each test but in our case it is a little different. Each time we need to start Katana and the docker container so it still takes a little time. So I would be in favor of grouping some of your tests together, for example:
   - If I'm not mistaken, your `test_mempool_pooled_transaction_hashes` test checks the hashes of transactions added to the pool against the expected. So we could make these assertions directly at the end of your `test_mempool_add_transactions` test. This would add additional assertions but it would limit the number of tests.
   - I think you can do the same thing to group several tests together as soon as possible.

**Comment by eugypalu** [comment]
> On the other hand, I know that it is rather recommended to separate the tests to have unit test as much as possible and limit the scope of each test but in our case it is a little different. Each time we need to start Katana and the docker container so it still takes a little time. So I would be in favor of grouping some of your tests together, for example:
> * If I'm not mistaken, your `test_mempool_pooled_transaction_hashes` test checks the hashes of transactions added to the pool against the expected. So we could make these assertions directly at the end of your `test_mempool_add_transactions` test. This would add additional assertions but it would limit the number of tests.
> * I think you can do the same thing to group several tests together as soon as possible.

Fixed, thanks

**Review by tcoratger — APPROVED** [review]

**Comment by tcoratger** [comment]
Nice, lgtm, pending https://github.com/kkrt-labs/kakarot-rpc/pull/1352 to be merged

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:359]: can be moved out of the for loop

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:371]: I think you can do `..Default::default()` for `input` and `access_list`.

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:380]: I think you can do `katana.eoa().address()` instead of recovering the signer. Also, since this will be the same each loop, you can just do it outside the for loop

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:92]: can we move this above the `remove_transaction` test? It's a little confusing when reading the test

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:191]: can't we replace all this by retrieving expected_hashes in a vec and received hashes in another vec?

Comment (greged93) [reply tests/tests/mempool.rs:191]: also I think it's nicer when the assert follows the actual function call directly, it's easier to follow the test.
So something like
```rust
    // Get pending transactions
    let hashes = eth_provider.mempool().unwrap().pending_transactions();
    // Check if the first pending transaction hash matches
    assert_eq!(hashes[0].hash(), &signed_transactions[0].hash());
    // Check if the second pending transaction hash matches
    assert_eq!(hashes[1].hash(), &signed_transactions[1].hash());
    // Ensure the number of pending transactions matches the expected count
    assert_eq!(hashes.len(), transaction_number);```

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:256]: same as above

File: tests/tests/mempool.rs
Comment (greged93) [inline tests/tests/mempool.rs:267]: we aren't really testing the subscription here right?

**Comment by tcoratger** [comment]
@eugypalu Here I think that you will have some conflicts to solve after the merge of https://github.com/kkrt-labs/kakarot-rpc/pull/1352#event-14148899607. Can you rebase? The main thing is that now you have to call `mempool()` and `send_raw_transaction()` from the `eth_client` and not from the `eth_provider` anymore.

File: tests/tests/mempool.rs
Comment (eugypalu) [reply tests/tests/mempool.rs]: fixed on all the tests, thanks

File: tests/tests/mempool.rs
Comment (eugypalu) [reply tests/tests/mempool.rs]: No, it wasn’t testing the subscription. I removed it, I don’t think it was useful. What do you think?

Comment (greged93) [reply tests/tests/mempool.rs]: I think it makes sense to remove if you weren't actually testing the subscription yes

**Comment by eugypalu** [comment]
> @eugypalu Here I think that you will have some conflicts to solve after the merge of [#1352 (comment)](https://github.com/kkrt-labs/kakarot-rpc/pull/1352#event-14148899607). Can you rebase? The main thing is that now you have to call `mempool()` and `send_raw_transaction()` from the `eth_client` and not from the `eth_provider` anymore.

Hey @tcoratger, I've rebased and fixed the conflicts 👍

**Review by greged93 — APPROVED** [review]

**Pull request merged**
